internal/pkg/agent/install/progress.go

```diff
@@ -15,15 +15,98 @@ import (
 	"golang.org/x/exp/rand"
 )
 
+// ProgressTrackerStep is a currently running step.
+//
+// A step can produce a sub-step that is a step that is part of another step.
+type ProgressTrackerStep interface {
+	// Succeeded step is done and successful.
+	Succeeded()
+	// Failed step has failed.
+	Failed()
+	// StepStart creates a new step.
+	StepStart(msg string) ProgressTrackerStep
+}
+
+type progressTrackerStep struct {
+	tracker *ProgressTracker
+	prefix  string
+
+	finalizeFunc func()
+
+	rootstep bool
+	substeps bool
+	step     *progressTrackerStep
+}
+
+func newProgressTrackerStep(tracker *ProgressTracker, prefix string, finalizeFunc func()) *progressTrackerStep {
+	return &progressTrackerStep{
+		tracker:      tracker,
+		prefix:       prefix,
+		finalizeFunc: finalizeFunc,
+	}
+}
+
+// Succeeded step is done and successful.
+func (pts *progressTrackerStep) Succeeded() {
+	prefix := " "
+	if pts.substeps {
+		prefix = pts.prefix + "   "
+	}
+	if !pts.rootstep {
+		pts.tracker.printf("%sDONE\n", prefix)
+	}
+	pts.finalizeFunc()
+}
+
+// Failed step has failed.
+func (pts *progressTrackerStep) Failed() {
+	prefix := " "
+	if pts.substeps {
+		prefix = pts.prefix + "   "
+	}
+	if !pts.rootstep {
+		pts.tracker.printf("%sFAILED\n", prefix)
+	}
+	pts.finalizeFunc()
+}
```

Comment:
[Suggestion]
you could make a function that takes a string (`DONE` or `FAILED`) and then `Succeeded` and `Failed` call it with the right string.
but the chance someone will need to change either `Succeeded` or `Failed` and forget the other is small, so it's realy up to you